chemprop/models/ffn.py

```diff
@@ -0,0 +1,350 @@
+                            dropout=dropout,
+                            activation=activation,
+                        ),
+                        activation,
+                    )
+                else:
+                    self.ffn = nn.Identity()
+            self.ffn_readout = build_ffn(
+                first_linear_dim=base_output_size,
+                hidden_size=hidden_size,
+                num_layers=1,
+                output_size=output_size,
+                dropout=dropout,
+                activation=activation,
+            )
+            self.weights_readout = build_ffn(
+                first_linear_dim=base_output_size,
+                hidden_size=hidden_size,
+                output_size=1,
+                num_layers=weights_ffn_num_layers,
+                dropout=dropout,
+                activation=activation,
+            )
+        else:
+            if shared_ffn:
+                self.ffn_readout = nn.Sequential(
+                    ffn_base,
+                    build_ffn(
+                        first_linear_dim=base_output_size,
+                        hidden_size=hidden_size,
+                        num_layers=1,
+                        output_size=output_size,
+                        dropout=dropout,
+                        activation=activation,
+                    ),
+                )
+            else:
+                self.ffn_readout = build_ffn(
+                    first_linear_dim=features_size,
+                    hidden_size=hidden_size,
+                    num_layers=num_layers,
+                    output_size=output_size,
+                    dropout=dropout,
+                    activation=activation,
+                )
+        self.constraint = constraint
+        self.ffn_type = ffn_type
+
+    def forward(
+        self,
+        input: Tuple[torch.Tensor, List, torch.Tensor, List, torch.Tensor],
+        constraints: torch.Tensor,
+        bond_types: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        Runs the :class:`FFNAtten` on input.
+        :param input: A tuple of atom and bond informations of each molecule.
+        :param constraints: A PyTorch tensor which applies constraint on atomic/bond properties.
+        :param bond_types: A PyTorch tensor storing bond types of each bond determined by RDKit molecules.
+        :return: The output of the :class:`FFNAtten`, a PyTorch tensor containing a list of property predictions.
+        """
+        a_hidden, a_scope, b_hidden, b_scope, b2br = input
+
+        if self.ffn_type == "atom":
+            hidden = a_hidden
+            scope = a_scope
+        elif self.ffn_type == "bond":
+            forward_bond = b_hidden[b2br[:, 0]]
+            backward_bond = b_hidden[b2br[:, 1]]
+            scope = [((start - 1) // 2, size // 2) for start, size in b_scope]
+        else:
+            raise ValueError(f"Unrecognized ffn_type of {self.ffn_type}.")
+
+        if constraints is not None:
```

Comment:
```suggestion
       if self.constraint:
```

Reply:
I think the broader question is the level of abstraction here. Your object model makes it such that the atom/bond-level prediction always passes through an `FFNAtten` module _regardless of whether it is constrained_. Logically, the `FFNAtten` module exists for the purposes of **constrained** output, so why are using it to make _unconstrained_ predictions?

Reply:
Added!